token.go

```diff
@@ -0,0 +1,226 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tspclient
+
+import (
+	"bytes"
+	"context"
+	"crypto"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"math/big"
+	"time"
+
+	"github.com/notaryproject/tspclient-go/internal/cms"
+	"github.com/notaryproject/tspclient-go/internal/hashutil"
+	"github.com/notaryproject/tspclient-go/internal/oid"
+)
+
+// SignedToken is a parsed timestamp token with signatures.
+type SignedToken cms.ParsedSignedData
+
+// ParseSignedToken parses ASN.1 BER-encoded structure to SignedToken
+// without verification.
+// Callers should invoke Verify to verify the content before comsumption.
+func ParseSignedToken(ctx context.Context, berData []byte) (*SignedToken, error) {
+	signed, err := cms.ParseSignedData(berData)
+	if err != nil {
+		return nil, err
+	}
+	if !oid.TSTInfo.Equal(signed.ContentType) {
+		return nil, fmt.Errorf("unexpected content type: %v", signed.ContentType)
+	}
+	return (*SignedToken)(signed), nil
+}
+
+// Verify verifies the signed token as CMS SignedData.
+// An empty list of KeyUsages in VerifyOptions implies ExtKeyUsageTimeStamping.
+// The `Intermediates` in the verify options will be ignored and
+// re-contrusted using the certificates in the parsed signed token.
+func (t *SignedToken) Verify(ctx context.Context, opts x509.VerifyOptions) ([]*x509.Certificate, error) {
+	if len(t.SignerInfos) == 0 {
+		return nil, SignedTokenVerificationError{Msg: "signerInfo not found"}
+	}
+	if len(opts.KeyUsages) == 0 {
+		opts.KeyUsages = []x509.ExtKeyUsage{x509.ExtKeyUsageTimeStamping}
+	}
+	intermediates := x509.NewCertPool()
+	for _, cert := range t.Certificates {
+		intermediates.AddCert(cert)
+	}
+	opts.Intermediates = intermediates
+	signed := (*cms.ParsedSignedData)(t)
+	var lastErr error
+	for _, signerInfo := range t.SignerInfos {
+		signingCertificate, err := t.GetSigningCertificate(&signerInfo)
+		if err != nil {
+			lastErr = SignedTokenVerificationError{Detail: err}
+			continue
+		}
+		certChain, err := signed.VerifySigner(ctx, &signerInfo, signingCertificate, opts)
+		if err != nil {
+			lastErr = SignedTokenVerificationError{Detail: err}
+			continue
+		}
+		// RFC 3161 2.3: The corresponding certificate MUST contain only one
+		// instance of the extended key usage field extension.
+		if len(signingCertificate.ExtKeyUsage) == 1 &&
+			signingCertificate.ExtKeyUsage[0] == x509.ExtKeyUsageTimeStamping &&
+			len(signingCertificate.UnknownExtKeyUsage) == 0 {
+			return certChain, nil
```

Comment:
Can we mention in the `Verify` method's comment that it returns as long as one signer info is valid?

Reply:
Added.